File: server/app/services/profile_engine.py
import json
from typing import Dict, Any, List
from langchain_openai import ChatOpenAI
from langchain_core.prompts import PromptTemplate
from langchain_core.output_parsers import JsonOutputParser
from app.core.config import settings
from app.schemas.profile import QuestionnaireSubmit

def get_llm():
    return ChatOpenAI(
        api_key=settings.DEEPSEEK_API_KEY,
        base_url=settings.DEEPSEEK_API_BASE,
        model="deepseek-chat",  # DeepSeek 的对话模型
        temperature=0.3, # 偏稳定输出
    )

class ProfileEngine:
    def __init__(self):
        self.llm = get_llm()

    async def generate_profile_data(self, answers: QuestionnaireSubmit) -> Dict[str, Any]:
        """
        利用大模型分析用户问卷，生成个性化的戒烟画像权重。
        """
        
        prompt_template = """
        你是一个专业的戒烟心理辅导专家。你需要分析用户填写的戒烟心理问卷，并生成标准化的 JSON 数据供系统调度使用。
        
        用户的问卷答案如下：
        {answers}
        
        请严格按照以下 JSON 格式输出你的分析结果，不要包含任何额外的说明文字：
        {{
            "smoking_years": 整数（根据年龄推算，如果是初中大概是10年以上，如果不好判断默认5）,
            "cigarettes_per_day": 浮点数（根据选项如"1-5支"取平均3）,
            "pack_price": 浮点数（根据选项如"15-25"取20）,
            "dependence_level": "low" | "medium" | "mediumHigh" | "high",
            "user_types": ["字符串数组，例如：工作休息触发型、社交需求型、情绪依赖型"],
            "motivation_weights": {{"字典格式，键必须是 health, money, baby, partner, parents, freedom, control 等，值是 0.0 到 1.0 之间的权重，总和不必为1"}},
            "trigger_weights": {{"字典格式，键必须是 wakeUp, afterMeal, workTired, stress, bored 等场景英文缩写，值是 0.0 到 1.0 的权重"}},
            "belief_tags": ["字符串数组，用户存在的错误信念标签，例如：放松幻觉、奖励幻觉、无聊补偿"],
            "recommended_modules": ["字符串数组，系统推荐开启的功能，只能从这些里面选：ai_rescue, money_bank, baby_account, work_mode, relapse_repair, cognitive_course"]
        }}
        """
        
        prompt = PromptTemplate(
            template=prompt_template,
            input_variables=["answers"],
        )
        
        # 将 Pydantic 对象转为美化的 JSON 字符串供 LLM 读取
        answers_str = json.dumps(answers.model_dump(), ensure_ascii=False, indent=2)
        
        parser = JsonOutputParser()
        chain = prompt | self.llm | parser
        
        result = await chain.ainvoke({"answers": answers_str})
        return result

File: app/lib/services/profile_engine.dart
import 'package:flutter/foundation.dart';

class ProfileEngine {
  static Map<String, dynamic> generateProfile(Map<String, dynamic> answers) {
    debugPrint('Calculating profile from answers: $answers');
    
    // 基础数据
    int smokingYears = 5;
    if (answers['start_age'] == '初中或更早') smokingYears = 15;
    if (answers['start_age'] == '高中') smokingYears = 10;
    if (answers['start_age'] == '大学') smokingYears = 8;
    if (answers['start_age'] == '工作以后') smokingYears = 5;
    if (answers['start_age'] == '最近几年') smokingYears = 2;

    int cigarettesPerDay = 3;
    if (answers['cigarettes_per_day'] == '1-5 支') cigarettesPerDay = 3;
    if (answers['cigarettes_per_day'] == '6-10 支') cigarettesPerDay = 8;
    if (answers['cigarettes_per_day'] == '11-15 支') cigarettesPerDay = 13;
    if (answers['cigarettes_per_day'] == '16-20 支（约一包）') cigarettesPerDay = 20;
    if (answers['cigarettes_per_day'] == '一包以上') cigarettesPerDay = 30;

    double packPrice = 20.0;
    if (answers['pack_price'] == '15 元以下') packPrice = 12.0;
    if (answers['pack_price'] == '15-25 元') packPrice = 20.0;
    if (answers['pack_price'] == '25-40 元') packPrice = 30.0;
    if (answers['pack_price'] == '40-60 元') packPrice = 50.0;
    if (answers['pack_price'] == '60 元以上') packPrice = 80.0;

    String dependenceLevel = 'medium';
    if (answers['first_cigarette'] == '5 分钟以内') dependenceLevel = 'high';
    else if (answers['first_cigarette'] == '5-30 分钟') dependenceLevel = 'mediumHigh';
    else if (answers['first_cigarette'] == '30 分钟到 1 小时') dependenceLevel = 'medium';
    else dependenceLevel = 'low';

    // 分析标签
    List<String> userTypes = [];
    if (answers['smoke_identity'] == '🧑‍🤝‍🧑 朋友' || answers['smoke_identity'] == '🤗 陪伴') userTypes.add('情感依赖型');
    if (answers['smoke_identity'] == '🤝 社交工具') userTypes.add('社交需求型');
    if (answers['smoke_identity'] == '🛋️ 放松' || answers['smoke_identity'] == '🎁 奖励') userTypes.add('缓解压力型');
    
    if (userTypes.isEmpty) userTypes.add('习惯依赖型');

    return {
      "smoking_years": smokingYears,
      "cigarettes_per_day": cigarettesPerDay,
      "pack_price": packPrice,
      "dependence_level": dependenceLevel,
      "user_types": userTypes,
      "motivation_weights": _extractList(answers['motivations']),
      "trigger_weights": _extractList(answers['triggers']),
      "belief_tags": _extractList(answers['previous_attempts']),
      "recommended_modules": ["ai_rescue", "money_bank", "health_recovery"]
    };
  }

  static List<String> _extractList(dynamic value) {
    if (value is List) {
      return value.map((e) => e.toString()).toList();
    }
    return [];
  }
}
